### DISCUSSION TITLE
Java - 2 - 성은, 일구, 호현

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#21] [질의 응답]
Java 질의응답

**Comment by seongeun42** [comment]
### ❓️추상 클래스와 인터페이스의 차이점에 대해서 말씀해주세요.

<br>

### 💬 답변

<blockquote>

* 추상클래스는 추상메소드를 하나 이상 가지는 클래스이다.
   * 그렇지 않은 경우도 있다. 하지만 일반적으로는 추상메소드를 포함한다.
   * 위 조건을 만족하면 다른 클래스와 동일하게 다양한 변수나 함수를 가질 수 있다.
   * 인스턴스를 생성할 수 없다.
   * extends를 통해 상속받는다.
   * 다중 상속 불가능하다.
* 인터페이스는 모든 메소드가 추상메소드여야 한다.
   * 단, 디폴트 메소드는 예외. 
   * 변수를 가질 수 없고 상수(public static final)만 가질 수 있으며, 모든 클래스와 상수의 접근제한자는 public이다.
   * 역시 인스턴스를 생성할 수 없다.
   * implements를 통해 상속받는다.
   * 다중 상속 가능하다.

</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️ 추상 클래스와 인터페이스의 사용 용도는 어떻게 다른가요?

<br>

### 💬 답변

<blockquote>

추상 클래스는 자식 클래스에서 공통적으로 사용할 부분은 미리 구현해놓고, 특정 부분은 오버라이딩하게 함으로써 공통 기능에 대한 코드의 중복을 줄이는 목적이 강하다. 따라서 클래스를 확장해나가는 느낌으로 사용한다. 
반면에 인터페이스는 자식 클래스에게 일종의 강제성을 강요한다. 즉, 인터페이스는 자식 클래스들이 구현해야할 메소드들의 정의를 명시하고, 그 메소드들을 구현토록 한다. 이를 통해 서로 다른 클래스가 특정 동작을 공통적으로 수행하도록 하고 싶을 때 사용할 수 있다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️객체 지향 프로그래밍의 4가지 특징에 대해서 말씀해주세요.

<br>

### 💬 답변

<blockquote>

* oop is A PIE
   * Abstraction: 추상화, 객체들의 공통적인 특성 (변수, 메소드 등)을 묶어 하나의 클래스로 표현한다.
   * Polymorphism: 다형성, 한 타입의 참조변수를 통해 여러 타입의 객체를 참조할 수 있도록 만든다.
   * Inheritance: 상속, 부모클래스의 변수 및 메소드를 자식 클래스에서 상속받아 사용할 수 있다.(다형성이랑 연관)
   * Encapsulation: 캡슐화, 클래스 내부의 데이터나 코드 등을 캡슐처럼 하나로 묶고, 외부에서 알 수 없게 접근을 제한하여 Loosely Coupled를 지향한다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️객체 지향 프로그래밍 5대 원칙에 대해 설명해주세요.

<br>

### 💬 답변

<blockquote>

**SOLID**

- SRP(단일 책임 원칙) : 하나의 클래스는 하나의 책임만 가져야 한다. 무언가 변경할 때, 그로 인한 파급 효과가 적을수록 잘 지킨 것. 하나의 클래스가 많은 책임을 가질수록 해당 클래스를 변경했을 때 다른 것도 함께 수정하게 되기 때문이다.

- OCP(개방-폐쇄 원칙) : 확장에는 열려 있고, 변경에는 닫혀 있어야 한다는 원칙. 다형성을 활용해 구현체는 바꾸더라도 다른 코드는 전혀 바꿀 필요 없는 일이 이에 해당한다.

- LSP(리스코프 치환 원칙) : 프로그램의 정확성을 깨뜨리지 않고 상위 타입의 인스턴스를 하위 타입의 인스턴스로 바꿀 수 있어야 한다는 원칙. 이 원칙을 지키지 않을 경우, 예를 들어 인터페이스에서 걷기를 두 다리로 걷기로 해놨는데, 하위 타입에서 두 팔로 걷기로 구현해버리면 문제가 생기게 된다.

- ISP(인터페이스 분리 원칙) : 인터페이스를 만들 때 하나로 뭉뚱그려 범용성 있는 인터페이스로 만드는 게 아니라, 인터페이스를 사용할 클라이언트를 기준으로 최대한 분리해서 만들어야 인터페이스가 명확해지고, 대체하기도 쉬워진다.

- DIP(의존 역전 원칙) : 구현체가 아닌 추상체인 인터페이스에 의존해야 한다. 인터페이스에 의존할 경우에는 다른 구현체로 바꿀 수 있지만, 구현체에 의존하면 다른 구현체로는 바꿀 수 없기 때문이다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️오버라이딩과 오버로딩에 대해 설명해주세요.

<br>

### 💬 답변

<blockquote>

* 오버로딩
   * 메소드의 이름이 같고, 매개변수의 타입이나 개수가 다른 메소드들을 추가적으로 정의하는 것이다.
   * 메소드의 리턴 타입만 다르거나, 변수의 이름만 다르고 그 타입들은 동일하면 오버로딩을 할 수 없다.
      * 즉, 매개변수의 차이로만 재정의 할 수 있다.
   * 접근제한자는 자유롭게 지정할 수 있다.
   * 다양한 데이터 타입에 대해 동일한 동작을 하는 메소드의 경우 오버로딩을 하기 적절한 메소드이다.
      * e.g. sum(int a, int b), sum(double a, double b)
* 오버라이딩
   * 부모클래스로부터 상속받은 메소드를 자식 클래스에서 재정의하는 것이다.
   * 메소드의 이름, 매개변수, 리턴타입 등이 모두 동일해야 한다. 
   * 자식 클래스에서 오버라이딩하는 메소드의 접근 제어자는 부모 클래스보다 더 좁게 설정할 수 없다.
   * 예외(Exception)는 부모 클래스의 메소드 보다 더 큰 예외를 throw할 수 없다.
      * e.g. 부모클래스에서 NullPointerException을 throw하는데 오버라이딩해서 Exception을 throw할 수 없다.
   * static메소드를 인스턴스의 메소드로, 또는 그 반대로 바꿀 수 없다.
   * 부모클래스의 메소드를 자식클래스에서 적절한 용도에 맞게 재정의해 사용할 때 오버라이딩을 활용할 수 있다.

</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️오버라이딩할 때 부모 클래스에서 정의된 static메소드를 자식 클래스에서 일반 메소드로, 또는 일반 메소드를 static 메소드로 바꿀 수 없는 이유에 대해 설명해주세요.

<br>

### 💬 답변

<blockquote>

* static 키워드가 붙는 경우에는 컴파일 시점에 객체가 생성되기 전에 메모리 영역에 올라가 결정.
* overriding의 경우에는 런타임 시점에 사용될 메소드를 결정.
* 따라서 부모클래스에 있는 static이 붙는 method를 자식클래스에서 overriding 하려는 경우, overriding이 되지 않고 hiding이 됨. static method는 클래스 수준에서 호출되기 때문에 자식클래스에서 overriding한다기 보다 오히려 별개로 존재하는 새로운 함수가 define되는 개념에 가까움. 결국 overriding의 개념이 적용되지 않는다는 뜻.
* 즉 호출되는 overriding 된 일반 method는 자식 버젼의 method이겠지만 static method의 경우 어디서(부모클래스인지, 자식클래스인지) 호출되는지에 따라 달라지는 현상이 발생함.
* 이런 상황에서 부모클래스를 상속할 때 자식클래스가 static method를 일반 method로 바꾸는게 가능해진다면, static method를 일반 method로 변환하여 overriding한 자식클래스와 그렇게 안한 자식클래스에서 해당 method를 호출할때 동작차이가 발생함.
* 만약 변환하여 overriding 하는 것이 가능하다면 아래에서
```
public Class Parent {
    public static void foo(){
        System.out.println("method of parent");
    }
}

public Class ChildA extends Parent {
    public static void foo() {
        System.out.println("method of child A");
    }
}

public Class ChildB extends Parent {
    public void foo() {
        System.out.println("method of child B");
    }
}

public static void main(String[] args) {
    Parent c1 = new ChildA();
    ChildA c2 = new ChildA();
    Parent c3 = new ChildB();
    c1.foo();
    c2.foo();
    c3.foo();
}
```
출력 :
```
method of parent
method of child A
??????????????????
```
3번 line은 compile시점에 결정된 "method of parent"를 출력할지 runtime시점에 결정된 "method of child B"를 출력할지 애매모호해지는 상황이 발생됨
이런 상황을 방지하기 위해 변환하여 오버라이딩 하는것을 허용하지 않는 것으로 이해됨.


</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️정적 바인딩과 동적 바인딩은 무엇이고, 어떤 장단점이 있는지 설명해주세요.

<br>

### 💬 답변

<blockquote>

바인딩이란 함수 호출 시 함수의 정의와 함수의 실제 구현부를 연결하는 것을 의미하며, 정적 바인딩과 동적 바인딩이 있다.

- 정적 바인딩은 컴파일 타임에 실행될 함수가 결정됨
  - 컴파일시 값이 확정돼 안정적이고 실행 효율이 좋음
  - 오버로딩, private, final, static이 붙은 메소드는 정적 바인딩 대상
 
- 동적 바인딩은 실행 타임에 실행될 함수가 결정됨
  - 메모리나 시간의 효율은 떨어지더라도 여러 상황에 유연하게 대처할 수 있음
  - 오버라이딩은 동적 바인딩 대상

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️String, StringBuilder, StringBuffer에 대해 설명해주세요.

<br>

### 💬 답변

<blockquote>

![img1 daumcdn](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/83163231-fc62-4b1e-8953-bd0ac4b36e57)

* 공통적으로 자바에서 문자열을 다룰 때 사용하는 클래스들이다.
   * String은 Immutable하고 StringBuilder와 StringBuffer는 Mutable하다.
   * StringBuilder는 동기화를 지원하지 않기 때문에 멀티스레드 환경에서 사용하기 적합하지 않다.
   * StringBuffer는 동기화를 지원하므로 멀티스레드 환경에 보다 적합하다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️String이 Immutable함으로써 얻을 수 있는 장점은 뭔가요?

<br>

### 💬 답변

<blockquote>

불변하기 때문에 여러 군데서 String 객체를 참조하더라도 수정이 일어나지 않아 안전하다. 또한 같은 값일 경우 같은 객체를 사용하기 때문에 메모리 절약과 속도 향상에 이점이 있다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️자바에서 스레드를 사용하기 위한 두 가지 방법에 대해 설명하고 차이점을 알려주세요.

<br>

### 💬 답변

<blockquote>

* Thread 클래스 상속하기
   * Thread 클래스를 상속하여 run() 메소드를 오버라이딩
   * 이후, 객체를 생성하고 start() 메소드를 호출해 직접 사용
* Runnable 구현하기
   * Runnable 인터페이스를 구현하여 run() 메소드를 오버라이딩
   * Thread 생성자 중 Runnable 인자형을 받는 생성자를 통해 Thread 객체를 만들고, start() 메소드를 호출해 사용
* 사견
   * Thread 상속 보단 Runnable 구현이 더 권장된다고 생각
      * 이유1. 클래스를 상속받으면 다중 상속이 안 되기 때문 - 이게 핵심
      * 이유2. Runnable을 람다식을 이용해 넘겨주면 코드가 더 간결해짐

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️자바에서의 멀티스레드 동기화에 대해 아는 데로 설명해주세요.

<br>

### 💬 답변

<blockquote>

* synchronized 키워드
   * 이 키워드가 적용된 코드 영역을 critical section으로 지정해준다.
   * lock을 활용해 동기화
   * 중괄호 블럭을 만들 수도 있고, 메소드에 붙여서 사용할 수도 있다.
   * https://docs.oracle.com/javase/tutorial/essential/concurrency/sync.html
 * Semaphore 클래스도 사용 가능
 * 내용이 너무 많아서 제가 아는 도큐먼트 링크 달게요
    * https://jenkov.com/tutorials/java-concurrency/index.html

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️자바의 스레드풀(Thread Pool)은 무엇이고 구조가 어떻게 되나요? 자바는 왜 스레드풀을 사용할까요?

<br>

### 💬 답변

<blockquote>

* 스레드풀이란 자바에서 스레드들을 관리하는 방법
* Producer-Consumer 컨셉을 생각하면 이해하기 쉬움
* 전체 구조
   * 스레드풀에는 여러 스레드들이 미리 생성된 상태로 대기하고 있음
      * 즉, 프로그램에서 생성 요청을 무한대로 한다고 해서 그만큼의 스레드가 생성되지 않음.
   * 어플리케이션은 멀티 스레드로 처리해야할 작업을 작업 큐에 넣는다.
   * 스레드풀에서 대기중인 스레드들은 큐에 있는 작업을 꺼내 실행한다.
* 스레드를 너무 많이 생성하거나, 생성-해제를 반복하는 것은 시간적으로나 메모리적으로나 오버헤드가 매우 큰 작업이다.
   * 따라서, 적정 개수의 스레드를 스레드풀에 미리 생성해놓고, 위와 같은 방식으로 작업을 처리하는 게 스레드
   * 스레드 생성-해제에 대한 오버헤드를 줄일 수 있다
* 추가로 스레드마다 동일한 수의 task를 할당하는 것보다 스레드 풀을 활용하여 스레드에 task를 할당할 때,  workload imbalance를 줄일 수 있다

![다운로드](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/ce42931a-d2fc-4c1d-9215-922f6da7ea7a)

</blockquote>

<br>
